### DISCUSSION TITLE
Static build for Windows UWP

### DISCUSSION THREAD

**Discussion by brabebhin** [libass/libass#609] [Q&A] [answered]
Hi,

This might sound a little stupid, but here we go.

I am trying to build this for UWP to link it into ffmpeg build for UWP, using MSVC.
However, I cannot find any guide on how to do this (in fact, compiling this library seems to be google's most well guarded secret).

Any hinds on where I can find some clues on this?

Thanks.

**Comment by astiob** [comment, marked as answer]
Hi!

An unofficial Visual Studio solution for libass exists in [ShiftMediaProject/libass](https://github.com/ShiftMediaProject/libass). It looks like it has WinRT/UWP support, too. It seems like it doesn’t track libass `master`, but it is up to date on the latest libass release, which may be sufficient for your needs.

A semi-official but much more outdated Visual Studio solution also exists in [libass/libass-msvc](https://github.com/libass/libass-msvc).

Another off-the-shelf alternative is to run Meson to produce an MSVC project for you using the tentative Meson configuration proposed in https://github.com/libass/libass/pull/330, as used in [TypesettingTools/Aegisub](https://github.com/TypesettingTools/Aegisub).

But if none of these suit you:

Currently, the only build environment libass trivially supports out of the box is a UNIX shell. For anything else, you basically need to emulate what libass’s `configure.ac` and `libass/Makefile.am` do. In any case, you also need to provide the libraries libass depends on: at minimum FreeType, FriBidi and HarfBuzz. (Fontconfig isn’t necessary on Windows Vista+, and libiconv isn’t needed for FFmpeg. And among HarfBuzz’s optional dependencies, none are necessary or useful for libass except Graphite2, which you may want to add for some advanced fonts but is not really necessary.)

The most basic solution is to take libass `master`, to create a project that contains all `.c` & `.h` files in the `libass` directory except `fontconfig` and `coretext` and that links to externally-built FreeType, FriBidi & HarfBuzz, and to add a `config.h` similar to [ShiftMediaProject’s](https://github.com/ShiftMediaProject/libass/blob/master/SMP/config.h) but without `CONFIG_ASM`, `CONFIG_FONTCONFIG` and `CONFIG_ICONV`. This will exclude optimized assembler code and thus make libass slower than it could be, but on the other hand this is easy. You do have to work out how to build FreeType, FriBidi and HarfBuzz though.

If you want to add assembler code (for x86/x64 only—not for ARM), I’m afraid I don’t know how it’s done myself, so you’ll have to see what ShiftMediaProject or libass-msvc do. In libass `master`, you’ll need to add `CONFIG_ASM` and also `ARCH_X86` to `config.h`.

Switching to a static build should be the same as always; just tell Visual Studio to do that. (Indeed, if you were to want a dynamic library, you’d additionally need to tell MSVC to export the symbols listed in `libass.sym`. But for a static build, that’s not necessary.)

Switching to UWP should be the same as elsewhere, too, but if all else fails, add the C macro definition `WINAPI_FAMILY=WINAPI_FAMILY_APP`.

**Reply by rcombs** [reply]
I'll also note that the usual solution to this is simply to build using gcc or clang in a POSIX shell. Since libass is a pure-C library, there are no ABI differences between builds made with different compilers, so a build made with gcc/clang can be linked against by code built with MSVC.

**Reply by astiob** [reply]
Right; very good point. The only issue is that building for UWP that way is more complicated. Maybe FFmpeg has a ready script for that or maybe MSYS2 has made this easier since I last checked, but I had to do [magic](https://github.com/libass/libass/blob/cbeea94fc70a6322ecce4c4ff30bef2a9028c18a/.github/workflows/ghci.yml#L124-L130) in our build script, which in turn I based on magic done in VLC’s build scripts and wouldn’t have come up with myself.

Then again, looking at that magic now, it seems link-time only, which shouldn’t affect a static build.

**Reply by astiob** [reply]
It’s been pointed out on IRC that if libass 0.15.2 or earlier is used (as opposed to `master`), then FreeType must be pinned to 2.11.0. With 2.11.1 or newer, libass will crash when encountering underline or strikeout.

**Reply by lukasf** [reply]
I am really looking forward for #330! Meson should be a lot easier to use than any of the other options. Especially, it can be easily integrated into an automatic build system.
